kdc/kerberos5.c

```diff
@@ -154,209 +154,271 @@ _kdc_is_anon_request(const KDC_REQ *req)
- * one, but preferring one that has default salt.
+ * Pick an enctype for a ticket session key or for a reply key that is in the
+ * intersection of:
+ *
+ *  - permitted_enctypes (local policy)
+ *  - requested enctypes (KDC-REQ-BODY's etype list)
+ *  - the principal's long-term keys' enctypes
+ *    or its configured etype list
  *
- * XXX This function does way way too much.  Split it up!
+ * There are two options
  *
- * XXX `etypes' and `len' are always `b->etype.val' and `b->etype.len' -- the
- *     etype list from the KDC-REQ-BODY, which is available here as
- *     `r->req->req_body', so we could just stop having it passed in.
+ *  - use local enctype preference (local policy), which will be "strongest is
+ *    most preferred"
+ *  - use the client's preference
  *
- * XXX Picking an enctype(s) for PA-ETYPE-INFO* is rather different than
- *     picking an enctype for a ticket's session key.  The former is what we do
- *     here when `(flags & KFE_IS_PREAUTH)', the latter otherwise.
+ * The first is clearly more secure, but has interop problems for clients that
+ * have multipler Kerberos implementations sharing the same credentials caches
+ * but having different sets of supported enctypes.
  */
-
 krb5_error_code
 _kdc_find_etype(astgs_request_t r, uint32_t flags,
-		krb5_enctype *etypes, unsigned len,
-		krb5_enctype *ret_enctype, Key **ret_key,
-		krb5_boolean *ret_default_salt)
+                krb5_enctype *ret_enctype)
 {
-    krb5_boolean use_strongest_session_key;
+    struct _krb5_encryption_type *et;
+    krb5_boolean use_strongest;
     krb5_boolean is_preauth = flags & KFE_IS_PREAUTH;
+    krb5_boolean use_client = flags & KFE_USE_CLIENT;
     krb5_boolean is_tgs = flags & KFE_IS_TGS;
     hdb_entry *princ;
-    krb5_principal request_princ;
-    krb5_error_code ret;
-    krb5_salt def_salt;
-    krb5_enctype enctype = ETYPE_NULL;
-    const krb5_enctype *p;
-    Key *key = NULL;
-    size_t i, k, m;
-
-    if (is_preauth && (flags & KFE_USE_CLIENT) &&
-        r->client->flags.synthetic)
+    const krb5_enctype *outer;
+    const krb5_enctype *inner;
+    size_t nouter, ninner;
+    int weak_allowed = 0;
+    int weak_offered = 0;
+    size_t o, i, k; /* iterators for outer, inner, keys */
+    size_t limit = 200; /* How hard we'll try */
+
+    /* Synthetic clients have no symmetric keys for PAC-ENC-* pre-auth */
+    if (is_preauth && use_client && r->client->flags.synthetic)
         return KRB5KDC_ERR_ETYPE_NOSUPP;
 
-    if ((flags & KFE_USE_CLIENT) && !r->client->flags.synthetic) {
-	princ = r->client;
-	request_princ = r->client_princ;
-    } else {
-	princ = r->server;
-	request_princ = r->server->principal;
+    if (!cached_kerberos_enctypes) {
+        /*
+         * NOTE: This is either all strong enctypes only, or all enctypes,
+         *       including weak ones, depending on configuration.
+         */
+        cached_kerberos_enctypes = krb5_kerberos_enctypes(r->context);
```

Comment:
Hmm, introduced global shared state when we've been trying to refactor everything to use `r`. Can you cache these in the context instead?

Reply:
It's a global in `krb5_kerberos_enctypes()`...

Reply:
Then why do we need to cache it? And why does it take a context argument? (Not in front of computer :))

Reply:
Because it's a pair of `static` variables inside that function, `krb5_kerberos_enctypes()`...

Reply:
Hmm, OK, so relying on `krb5_enctype_valid()` being invariant across contexts, which I guess is safe. But, why can't we just call `krb5_enctype_valid()` each time?

Reply:
Ah, that: because it does a lot of work on error that we don't need to do in this context.  I'm manually inlining just the piece we need.